whelktool/scripts/cleanups/2023/12/fmt-210-restore-primary-contributors.groovy

```diff
@@ -0,0 +1,77 @@
+        'https://id.kb.se/relator/woodEngraver',
+        'https://id.kb.se/relator/electrotyper'
+]
+        .collect { ['@id': it] } as Set
+
+def scriptId = "https://libris.kb.se/sys/globalchanges/2023/05/lxl-2512-move-contribution-by-relator-domain/script.groovy"
+
+selectBySqlWhere("collection = 'bib' and deleted = false and data#>>'{@graph,1,contribution}' is not null") { bib ->
+    def id = bib.doc.shortId
+    def instance = bib.graph[1]
+    def work = instance.instanceOf
+
+    if (!work || work['@id']) return
+
+    def newestToOldestVersion
+    try {
+        newestToOldestVersion = bib.getVersions().reverse()
+    } catch (Exception e) {
+        println(e.getStackTrace())
+        return
+    }
+    def versionTouchedByScriptIdx = newestToOldestVersion.findIndexOf { it.data['@graph'][0]['generationProcess']?['@id'] == scriptId }
+
+    if (versionTouchedByScriptIdx == -1) return
+
+    def versionBefore = newestToOldestVersion[versionTouchedByScriptIdx + 1]
+    def workContributionBefore = versionBefore.data['@graph'][1]['instanceOf']['contribution']
+    def primaryBefore = workContributionBefore.find { it['@type'] == 'PrimaryContribution' }
+
+    if (asList(primaryBefore?['role']).intersect(roles).isEmpty()) return
+
+    // Reaching here means that one of the concerned roles was moved from PrimaryContribution on work to "just" Contribution on instance by lxl-2512.
+
+    def currentPrimary = work['contribution']?.find { it['@type'] == 'PrimaryContribution' && it['agent'] == primaryBefore['agent'] }
+
+    // Decide which role should be in PrimaryContribution based on work type.
+    def newPrimaryRole = work['@type'] == 'StillImage'
+            ? ['@id': 'https://id.kb.se/relator/artist']
+            : (work['@type'] == 'Cartography' ? ['@id': 'https://id.kb.se/relator/cartographer'] : null)
+
+    def roleShort = { it['@id'].replaceAll(".*/", "") }
+
+    if (!newPrimaryRole) {
+        // Unable to decide which role to put in PrimaryContribution
+        unhandled.println([id, work['@type'], asList(currentPrimary?['role']).collect(roleShort)].join('\t'))
+        return
+    }
+
+    if (currentPrimary) {
+        // The agent is still present in work PrimaryContribution
+        if (!asList(currentPrimary['role']).contains(newPrimaryRole)) {
+            // The current role(s) is neither artist nor cartographer, report.
+            unhandled.println([id, work['@type'], asList(currentPrimary['role']).collect(roleShort)].join('\t'))
+        }
+    } else {
+        // The agent is no longer present in work PrimaryContribution, readd it with relevant role.
+        def newPrimary = ['@type': 'PrimaryContribution', 'agent': primaryBefore['agent'], 'role': [newPrimaryRole]]
+        work['contribution'] = [newPrimary] + asList(work['contribution'])
+        addedPrimary.println([id, newPrimary['agent'], newPrimary['role'].collect(roleShort)].join('\t'))
+        bib.scheduleSave()
+    }
```

Comment:
Don't we have to check if there is currently _another_ `PrimaryContribution` agent on the work?

Reply:
I don't think that's necessary since if there is another PrimaryContribution it also has different agent (see line 49). Maybe there shouldn't be two PrimaryContribution ideally but since that was the case before I think it's ok to revert to that state.

Reply:
I don't think that's necessary since if there is another PrimaryContribution it also has different agent (see line 49). Maybe there shouldn't be two PrimaryContribution ideally but since that was the case before I think it's ok to revert to that state.